File: platforms/ios/Dezel/Dezel/Text/TextLayout.swift
import CoreText

/**
 * Break up text into lines that fits into specified bounds.
 * @class TextLayout
 * @since 0.5.0
 */
open class TextLayout: NSObject {

	//--------------------------------------------------------------------------
	// MARK: Properties
	//--------------------------------------------------------------------------

	/**
	 * The text layout's text.
	 * @property text
	 * @since 0.5.0
	 */
	open var text: NSAttributedString = NSAttributedString(string: "") {
		willSet {
			self.invalidate()
		}
	}

	/**
	 * The text layout's ellipsis.
	 * @property ellipsis
	 * @since 0.5.0
	 */
	open var ellipsis: NSAttributedString = NSAttributedString(string: "…") {
		willSet {
			self.invalidate()
		}
	}

	/**
	 * The text layer's text kerning.
	 * @property textKerning
	 * @since 0.1.0
	 */
	open var textKerning: CGFloat = 0 {
		willSet {
			self.invalidate()
		}
	}

	/**
	 * The text layer's text leading.
	 * @property textLeading
	 * @since 0.1.0
	 */
	open var textLeading: CGFloat = 0 {
		willSet {
			self.invalidate()
		}
	}

	/**
	 * The text layer's text baseline.
	 * @property textBaseline
	 * @since 0.1.0
	 */
	open var textBaseline: CGFloat = 0 {
	 	willSet {
	 		self.invalidate()
		}
	}

	/**
	 * The text layer's text placement.
	 * @property textPlacement
	 * @since 0.1.0
	 */
	open var textPlacement: TextPlacement = .middle {
		willSet {
			self.invalidate()
		}
	}

	/**
	 * The text layout's text overflow.
	 * @property textOverflow
	 * @since 0.5.0
	 */
	open var textOverflow: TextOverflow = .ellipsis {
		willSet {
			self.invalidate()
		}
	}

	/**
	 * The text layout's line limit.
	 * @property lines
	 * @since 0.5.0
	 */
	open var lines: Int = 0 {
		willSet {
			self.invalidate()
		}
	}

	/**
	 * The text layout given size.
	 * @property bounds
	 * @since 0.5.0
	 */
	private(set) public var bounds: CGSize = .zero

	/**
	 * The text layout's computed limits.
	 * @property extent
	 * @since 0.5.0
	 */
	private(set) public var limits: CGSize = .zero

	/**
	 * The text layout's size that is taken.
	 * @property extent
	 * @since 0.5.0
	 */
	private(set) public var extent: CGSize = .zero

	/**
	 * The text layout's size measured by the frame setter.
	 * @property length
	 * @since 0.5.0
	 */
	private(set) public var length: CGSize = .zero

	/**
	 * The text layout's frame offset.
	 * @property offset
	 * @since 0.5.0
	 */
	private(set) public var offset: CGPoint = .zero

	/**
	 * Whether the text layout is invalid.
	 * @property invalid
	 * @since 0.5.0
	 */
	private(set) public var invalid: Bool = false

	/**
	 * @property setter
	 * @since 0.5.0
	 * @hidden
	 */
	private var setter: CTFramesetter!

	/**
	 * @property frame
	 * @since 0.5.0
	 * @hidden
	 */
	private var frame: CTFrame!

	//--------------------------------------------------------------------------
	// MARK: Methods
	//--------------------------------------------------------------------------

	/**
	 * Returns the character at the specified location.
	 * @method find
	 * @since 0.5.0
	 */
	open func find(at point: CGPoint) -> NSAttributedString? {

		if (self.text.length == 0) {
			return nil
		}

		guard let lines = CTFrameGetLines(self.frame) as? [CTLine] else {
			return nil
		}

		let limit = self.getLineLimit()

		let bounds = self.bounds
		let extent = self.extent

		var offset = self.offset.y
		offset -= bounds.height
		offset += extent.height

		switch (self.textPlacement) {

			case .middle:
				offset += self.extent.alignMiddle(of: bounds).y
			case .bottom:
				offset += self.extent.alignBottom(of: bounds).y

			default: break
		}

		offset = offset + self.textBaseline
		offset = offset * -1

		var transform = CGAffineTransform.identity
		transform = transform.translatedBy(x: 0, y: bounds.height)
		transform = transform.scaledBy(x: 1, y: -1)
		transform = transform.translatedBy(x: 0, y: offset)
		let point = point.applying(transform)

		let offsets = CTFrameGetLineOffsets(frame, lines.count)

		for (index, line) in lines.enumerated() {

			var bounds = CTLineGetBoundsWithOptions(line, .useOpticalBounds)
			bounds.origin.x = offsets[index].x
			bounds.origin.y = offsets[index].y

			if (point.x < bounds.minX || point.x > bounds.maxX ||
				point.y < bounds.minY || point.y > bounds.maxY) {
				continue
			}

			let index = CTLineGetStringIndexForPosition(line, CGPoint(x: point.x - bounds.origin.x, y: point.y))
			if (index != kCFNotFound) {
				return self.text.substring(offset: index == self.text.length ? self.text.length - 1 : index, length: 1)
			}

			if (index == limit) {
				break
			}
		}

		return nil
	}

	/**
	 * Invalidates the layout.
	 * @method invalidate
	 * @since 0.5.0
	 */
	open func invalidate() {
		self.invalid = true
	}

	/**
	 * Builds the layout.
	 * @method build
	 * @since 0.5.0
	 */
	open func build(in bounds: CGSize) {

		if (self.text.length == 0) {
			return
		}

		if (self.invalid == false) {
			if (self.bounds.equalTo(bounds) == false &&
				self.extent.equalTo(bounds) == false) {
				self.invalid = true
			}
		}

		self.bounds = bounds

		if (self.invalid == false) {
			return
		}

		var w = bounds.width
		var h = bounds.height

		if (w == 0) { w = CGFloat.max }
		if (h == 0) { h = CGFloat.max }

		let limits = CGSize(width: w, height: h)

		self.limits = limits
		self.extent = .zero
		self.length = .zero
		self.offset = .zero

		self.frame = CTFrameCreateWithAttributedString(self.text, size: limits)

		guard let lines = CTFrameGetLines(self.frame) as? [CTLine] else {
			return
		}

		let limit = self.getLineLimit()

		for (index, line) in lines.enumerated() {

			let bounds = CTLineGetBoundsWithOptions(line, [])

			let w = bounds.width
			let h = bounds.height

			if (self.length.width < w) {
				self.length.width = w
			}

			if (index <= limit) {

				if (self.extent.width < w) {
					self.extent.width = w
				}

				self.extent.height += ceil(h)
				self.extent.height += self.textLeading
			}
		}

		self.length.height = kTextLayoutFrameHeight
		self.length.ceil()
		self.extent.ceil()

		self.offset.x = self.length.width - self.extent.width
		self.offset.y = self.length.height - self.extent.height

		if (self.extent.width < self.bounds.width) {
			self.extent.width = self.bounds.width
		}

		self.invalid = false
	}

	/**
	 * Draws the text layout.
	 * @method invalidate
	 * @since 0.5.0
	 */
	open func draw(in context: CGContext) {

		guard let lines = CTFrameGetLines(self.frame) as? [CTLine] else {
			return
		}

		let limit = self.getLineLimit()

		var height = CGFloat(0)
		let bounds = self.bounds
		let extent = self.extent

		var offset = self.offset.y
		offset -= bounds.height
		offset += extent.height

		switch (self.textPlacement) {

			case .middle:
				offset += extent.alignMiddle(of: bounds).y
			case .bottom:
				offset += extent.alignBottom(of: bounds).y

			default: break
		}

		offset = offset + self.textBaseline
		offset = offset * -1

		context.translateBy(x: 0, y: bounds.height)
		context.scaleBy(x: 1, y: -1)
		context.translateBy(x: 0, y: offset)

		let offsets = CTFrameGetLineOffsets(self.frame, lines.count)

		for (index, value) in lines.enumerated() {

			var line = value

			height += self.measureHeight(line)

			context.textPosition.x = offsets[index].x
			context.textPosition.y = offsets[index].y

			let finished = self.isLastVisibleLine(
				line,
				lines: lines,
				index: index,
				limit: limit,
				height: height
			)

			if (finished) {

				if (self.textOverflow == .ellipsis && index < lines.count - 1) {

					if let last = self.text.truncate(range: CTLineGetStringRange(line), bounds: self.bounds, character: self.ellipsis) {

						if (context.textPosition.x > 0) {

							/*
							 * Reposition the truncated line using the ratio of new empty
							 * space versus the old empty space
							 */

							let x = (
								(self.bounds.width - CTLineGetBoundsWithOptions(last, .useOpticalBounds).size.width) /
								(self.bounds.width - CTLineGetBoundsWithOptions(line, .useOpticalBounds).size.width)
							) * context.textPosition.x

							context.textPosition.x = CGFloat.sround(x)
						}

						line = last
					}
				}
			}

			CTLineDraw(line, context)

			if (finished) {
				break
			}
		}
	}

	//--------------------------------------------------------------------------
	// MARK: Private API
	//--------------------------------------------------------------------------

	/**
	 * @method measureHeight
	 * @since 0.5.0
	 * @hidden
	 */
	private func measureHeight(_ line: CTLine) -> CGFloat {
		return CTLineGetBoundsWithOptions(line, .useOpticalBounds).height + self.textLeading
	}

	/**
	 * @method isLastVisibleLine
	 * @since 0.5.0
	 * @hidden
	 */
	private func isLastVisibleLine(_ line: CTLine, lines: [CTLine], index: Int, limit: Int, height: CGFloat) -> Bool {
		return index == limit || (index < lines.count - 1 && self.measureHeight(line) + height > self.bounds.height)
	}

	/**
	 * @method getLineLimit
	 * @since 0.5.0
	 * @hidden
	 */
	private func getLineLimit() -> Int {
		return self.lines > 0 ? self.lines - 1 : CFArrayGetCount(CTFrameGetLines(self.frame)) - 1
	}
}

/**
 * @const kTextLayoutFrameHeight
 * @since 0.5.0
 * @hidden
 */
internal let kTextLayoutFrameWidth = CGFloat(10000)
internal let kTextLayoutFrameHeight = CGFloat(10000)

File: platforms/android/Dezel/dezel/src/main/java/ca/logaritm/dezel/text/TextLayout.kt
package ca.logaritm.dezel.text

import android.graphics.Canvas
import android.graphics.Rect
import android.text.Spanned
import android.text.StaticLayout
import android.text.TextPaint
import android.util.SizeF
import ca.logaritm.dezel.extension.Delegates
import ca.logaritm.dezel.extension.text.SpannableString
import ca.logaritm.dezel.extension.type.max
import ca.logaritm.dezel.util.geom.Size
import ca.logaritm.dezel.text.layout.StaticLayoutBuilder
import ca.logaritm.dezel.view.type.TextLocation
import ca.logaritm.dezel.view.type.TextOverflow

/**
 * @class TextLayout
 * @since 0.5.0
 */
open class TextLayout {

	//--------------------------------------------------------------------------
	// Properties
	//--------------------------------------------------------------------------

	/**
	 * @property text
	 * @since 0.5.0
	 */
	open var text: Spanned by Delegates.OnSet(SpannableString("") as Spanned) {
		this.invalidate()
	}

	/**
	 * @property textKerning
	 * @since 0.5.0
	 */
	open var textKerning: Float by Delegates.OnSet(0f) {
		this.invalidate()
	}

	/**
	 * @property textLeading
	 * @since 0.5.0
	 */
	open var textLeading: Float by Delegates.OnSet(0f) {
		this.invalidate()
	}

	/**
	 * @property textBaseline
	 * @since 0.5.0
	 */
	open var textBaseline: Float by Delegates.OnSet(0f) {
		this.invalidate()
	}

	/**
	 * @property textLocation
	 * @since 0.7.0
	 */
	open var textLocation: TextLocation by Delegates.OnSet(TextLocation.MIDDLE) {
		this.invalidate()
	}

	/**
	 * @property textOverflow
	 * @since 0.5.0
	 */
	open var textOverflow: TextOverflow by Delegates.OnSet(TextOverflow.ELLIPSIS) {
		this.invalidate()
	}

	/**
	 * @property maxLines
	 * @since 0.7.0
	 */
	open var maxLines: Int by Delegates.OnSet(0) {
		this.invalidate()
	}

	/**
	 * @property bounds
	 * @since 0.5.0
	 */
	public var bounds: Size = Size()
		private set

	/**
	 * @property limits
	 * @since 0.5.0
	 */
	public var limits: Size = Size()
		private set

	/**
	 * @property extent
	 * @since 0.5.0
	 */
	public var extent: Size = Size()
		private set

	/**
	 * @property invalid
	 * @since 0.5.0
	 */
	public var invalid: Boolean = false
		private set

	/**
	 * @property layout
	 * @since 0.5.0
	 * @hidden
	 */
	private lateinit var layout: StaticLayout

	/**
	 * @property frame
	 * @since 0.5.0
	 * @hidden
	 */
	private var frame: Rect = Rect()

	//--------------------------------------------------------------------------
	// Methods
	//--------------------------------------------------------------------------

	/**
	 * @method build
	 * @since 0.5.0
	 */
	open fun find(x: Float, y: Float): Spanned? {

		if (this.text.length == 0) {
			return null
		}

		val line = this.layout.getLineForVertical(y.toInt())
		if (line == -1) {
			return null
		}

		var index = this.layout.getOffsetForHorizontal(line, x)
		if (index == this.text.length) {
			index = this.text.length - 1
		}

		return this.text.subSequence(index, index + 1) as Spanned
	}

	/**
	 * @method build
	 * @since 0.5.0
	 */
	open fun build(bounds: SizeF, paint: TextPaint) {
		this.build(Size(bounds), paint)
	}

	/**
	 * @method build
	 * @since 0.5.0
	 */
	open fun build(bounds: Size, paint: TextPaint) {

		if (this.text.length == 0) {
			return
		}

		if (this.invalid == false) {
			if (this.bounds.equals(bounds) == false &&
				this.extent.equals(bounds) == false) {
				this.invalid = true
			}
		}

		this.bounds = Size(bounds)

		if (this.invalid == false) {
			return
		}

		var w = bounds.width
		var h = bounds.height

		if (w == 0f) w = Float.max
		if (h == 0f) h = Float.max

		val limits = Size(w, h)

		this.limits = limits
		this.extent = Size()

		this.layout = StaticLayoutBuilder.build(
			paint,
			this.bounds,
			this.limits,
			this.text,
			this.textLeading,
			this.textKerning,
			this.textOverflow,
			this.maxLines
		)

		val last = this.getLastLine()

		for (line in 0 until this.layout.lineCount) {

			val width = this.layout.getLineWidth(line)

			if (this.extent.width < width) {
				this.extent.width = width
			}

			this.extent.height = this.layout.getLineBottom(line).toFloat()

			if (line == last) {
				break
			}
		}

		if (this.extent.width < this.bounds.width) {
			this.extent.width = this.bounds.width
		}

		this.extent.ceil()

		this.invalid = false
	}

	/**
	 * @method invalidate
	 * @since 0.5.0
	 */
	open fun invalidate() {
		this.invalid = true
	}

	/**
	 * @method draw
	 * @since 0.5.0
	 */
	open fun draw(canvas: Canvas) {

		val bounds = this.bounds
		val extent = this.extent

		var offset = 0f

		when (this.textLocation) {

			TextLocation.MIDDLE -> offset += extent.toMiddleOf(bounds).y
			TextLocation.BOTTOM -> offset += extent.toBottomOf(bounds).y
			TextLocation.TOP    -> {}

		}

		offset += this.textBaseline

		this.frame.set(
			0, 0,
			this.bounds.width.toInt(),
			this.bounds.height.toInt()
		)

		canvas.save()
		canvas.clipRect(this.frame)
		canvas.translate(0f, offset)

		this.layout.draw(canvas)

		canvas.restore()
	}

	//--------------------------------------------------------------------------
	// Private API
	//--------------------------------------------------------------------------

	private fun getLastLine(): Int {
		return this.maxLines - 1
	}
}